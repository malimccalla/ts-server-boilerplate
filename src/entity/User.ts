import * as bcrypt from 'bcryptjs';
import {
  BaseEntity,
  BeforeInsert,
  Column,
  Entity,
  PrimaryGeneratedColumn,
} from 'typeorm';

@Entity('users')
export class User extends BaseEntity {
  @PrimaryGeneratedColumn('uuid') id: string;

  @Column('varchar', { length: 255 })
  email: string;

  @Column('text') password: string;

  @Column('boolean', { default: false })
  confirmed: boolean;

  @Column('boolean', { default: false })
  locked: boolean;

  @BeforeInsert()
  async hashPassword() {
    const password = await bcrypt.hash(this.password, 10);
    this.password = password;
  }
}
