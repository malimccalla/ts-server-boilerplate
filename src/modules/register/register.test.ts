import * as faker from 'faker';
import { Connection } from 'typeorm';

import { User } from '../../entity/User';
import { createTestConn } from '../../test/createTestConn';
import { TestClient } from '../../test/TestClient';

faker.seed(Date.now() + Math.random());

describe('Register', () => {
  const password = faker.internet.password(10);

  describe('Happy path', () => {
    let conn: Connection;

    beforeEach(async () => {
      conn = await createTestConn();
    });

    afterEach(async () => {
      await conn.close();
    });

    test('should return a user on creation', async () => {
      const client = new TestClient();
      const email = faker.internet.email();

      const res = await client.register({ email, password });

      const user = await User.findOne({ where: { email } });

      expect(res.register.ok).toBe(true);
      expect(res.register.user).toBeTruthy();
      expect(res.register.user!.email).toEqual(email);

      // check password has been hashed
      expect(user!.password).not.toEqual(password);
    });

    test('user should not be confirmed on creation', async () => {
      const email = faker.internet.email();

      const client = new TestClient();
      await client.register({ email, password });

      const user = await User.findOne({ where: { email } });
      expect(user!.confirmed).toBe(false);
    });
  });

  describe('Email validation', () => {
    let conn: Connection;

    beforeEach(async () => {
      conn = await createTestConn();
    });

    afterEach(async () => {
      await conn.close();
    });

    test('should not allow an invalid email', async () => {
      const client = new TestClient();
      const email = 'Master Blaster';

      const res = await client.register({ email, password });

      expect(res.register.ok).toBe(false);
      expect(res.register.errors).toHaveLength(1);
      expect(res.register.user).toBeFalsy();
      expect(res.register.errors).toMatchSnapshot();
    });

    test('should not allow duplicate emails', async () => {
      const client = new TestClient();
      const email = faker.internet.email();

      await client.register({ email, password });
      const res = await client.register({ email, password });

      expect(res.register.ok).toBe(false);
      expect(res.register.errors).toHaveLength(1);
      expect(res.register.errors).toMatchSnapshot();

      expect(res.register.errors[0].path).toEqual('email');
      expect(res.register.errors[0].message).toEqual('Email already in use');
    });
  });

  describe('Password validation', () => {
    let conn: Connection;

    beforeEach(async () => {
      conn = await createTestConn();
    });

    afterEach(async () => {
      await conn.close();
    });

    test('should not allow a password less than 8 characters', async () => {
      const client = new TestClient();
      const email = faker.internet.email();

      const res = await client.register({ email, password: 'Hey Joe' });

      expect(res.register.ok).toBe(false);
      expect(res.register.errors).toHaveLength(1);
      expect(res.register.user).toBeFalsy();
      expect(res.register.errors).toMatchSnapshot();
    });
  });
});
