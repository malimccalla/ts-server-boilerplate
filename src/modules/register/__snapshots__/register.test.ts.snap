// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Email validation should not allow an invalid email 1`] = `
Array [
  Object {
    "message": "Invalid email",
    "path": "email",
  },
]
`;

exports[`Register Email validation should not allow duplicate emails 1`] = `
Array [
  Object {
    "message": "Email already in use",
    "path": "email",
  },
]
`;

exports[`Register Password validation should not allow a password less than 8 characters 1`] = `
Array [
  Object {
    "message": "Password must be longer than 8 characters",
    "path": "password",
  },
]
`;
